### PR TITLE
ARROW-6435: [Python] Use pandas null coding consistently on List and Struct types

### PR DESCRIPTION
**Pull request by wesm**
also fixes ARROW-6434

### CODE REVIEW

File: cpp/src/arrow/python/python_to_arrow.cc [hunk 1]
```diff
@@ -626,8 +644,82 @@ class ListConverter : public TypedConverter<TypeClass, ListConverter<TypeClass>>
   }
 
   template <int NUMPY_TYPE, typename Type>
-  Status AppendNdarrayTypedItem(PyArrayObject* arr);
-  Status AppendNdarrayItem(PyObject* arr);
+  Status AppendNdarrayTypedItem(PyArrayObject* arr) {
```
Comment (wesm) [inline cpp/src/arrow/python/python_to_arrow.cc]: Moved code. Unchanged

File: cpp/src/arrow/python/python_to_arrow.cc [hunk 2]
```diff
@@ -607,8 +607,26 @@ class StringConverter
...
+      return value_converter_->AppendMultiple(obj, value_length); \
+    }                                                             \
+    return AppendNdarrayTypedItem<NUMPY_TYPE, ArrowType>(arr);    \
+  }
+
+// Use internal::VisitSequence, fast for NPY_OBJECT but slower otherwise
+#define LIST_SLOW_CASE(TYPE)                                    \
+  case Type::TYPE: {                                            \
+    return value_converter_->AppendMultiple(obj, value_length); \
+  }
```
Comment (wesm) [inline cpp/src/arrow/python/python_to_arrow.cc]: Moved code

File: cpp/src/arrow/python/python_to_arrow.cc [hunk 3]
```diff
@@ -607,8 +607,26 @@ class StringConverter
...
+
+template <typename TypeClass, NullCoding null_coding>
+class ListConverter
+    : public TypedConverter<TypeClass, ListConverter<TypeClass, null_coding>,
+                            null_coding> {
```
Comment (wesm) [inline cpp/src/arrow/python/python_to_arrow.cc]: Relevant changes

File: cpp/src/arrow/python/python_to_arrow.cc [hunk 4]
```diff
@@ -959,16 +957,36 @@ Status GetConverter(const std::shared_ptr<DataType>& type, bool from_pandas,
...
-      *out = std::unique_ptr<SeqConverter>(
-          new StructConverter(from_pandas, strict_conversions));
+      if (from_pandas) {
+        *out = std::unique_ptr<SeqConverter>(
+            new StructConverter<NullCoding::PANDAS_SENTINELS>(from_pandas,
+                                                              strict_conversions));
+      } else {
+        *out = std::unique_ptr<SeqConverter>(
+            new StructConverter<NullCoding::NONE_ONLY>(from_pandas, strict_conversions));
+      }
```
Comment (wesm) [inline cpp/src/arrow/python/python_to_arrow.cc]: Relevant changes

**Comment by wesm** [comment]
There's another failure here, that separately came up in #5299. I'll add some tests and close #5299

**Comment by wesm** [comment]
OK, the Dask integration tests are passing again. Checking HDFS

**Comment by wesm** [comment]
OK, HDFS docker-compose job passing also now

**Comment by kszucs** [comment]
@ursabot crossbow test docker-hdfs-integration docker-dask-integration

**Comment by kszucs** [comment]
@ursabot crossbow test docker-hdfs-integration

**Comment by kszucs** [comment]
The HDFS build was cancelled, don't know why.

**Comment by wesm** [comment]
Rebased this and fixed the conflict from #5299 (which should not have been merged)

**Comment by wesm** [comment]
+1 

Travis CI: https://travis-ci.org/wesm/arrow/builds/581726782
Appveyor: https://ci.appveyor.com/project/wesm/arrow/builds/27234814

**Pull request closed without merging**